### ISSUE TITLE
which ensemble technique is used in auto sklearn

### ISSUE THREAD

**Issue by shabir1** [automl/auto-sklearn#1113]
which ensemble technique is used in auto sklearn, Is it Voting or any other technique?

**Comment by bitsbuffer** [comment]
It can compute Ensemble based on a greedy algorithm from Caruana paper.
http://www.cs.cornell.edu/~alexn/papers/shotgun.icml04.revised.rev2.pdf

**Comment by shabir1** [comment]
@bitsbuffer Thank You.

**Comment by shabir1** [comment]
I have another doubt.
I set `ensemble_size=10` and run the auto sklearn for 1 hour, `model.cv_results_` contains 243 rows out that 243 shows success status and 3 shows timeout. but when I print `model.show_models()` it prints only 3 models. How would I know which are those 3 models in `cv_results_`. I tried to sort `cv_results_` on `mean_test_score` key but the top 3 configs of `cv_results__` are not matching with `show_models()` results. And If I set `ensemble_size=10` why I got only 3 models from `show_models()`.
Below are the top 5 configs when sort cv_results_ on mean_test_score.
`adaboost` is not present in `show_models()`

mean_test_score | mean_fit_time | param_classifier:__choice__ | rank_test_scores | status | budgets | param_balancing:strategy | param_classifier:__choice__
-- | -- | -- | -- | -- | -- | -- | --
0.898388227049755 | 3.68728971481323 | extra_trees | 1 | Success | 0 | weighting | extra_trees
0.896986685353889 | 6.91696286201477 | adaboost | 2 | Success | 0 | none | adaboost
0.894183601962158 | 3.38019490242004 | extra_trees | 3 | Success | 0 | none | extra_trees
0.893482831114226 | 3.80625748634338 | extra_trees | 4 | Success | 0 | weighting | extra_trees
0.893482831114226 | 3.25829768180847 | extra_trees | 4 | Success | 0 | none | extra_trees

**Comment by mfeurer** [comment]
> And If I set ensemble_size=10 why I got only 3 models from show_models()

We use ensemble selection with repetitions, that means a model can be added multiple times to the ensemble, giving it a higher weight. `ensemble_size=10` means that there are 10 slots, and it turns out they are filled with 3 models in your case. Also, the ensemble does not contain the best configurations, but is constructed to optimize the given loss function on a holdout set, thus it can leave out the 3rd best model.

> How would I know which are those 3 models in cv_results_?

Unfortunately, this is not possible yet in an automated fashion and you would have to compare the hyperparameters by hand. It would indeed be great to have this information available as well.

**Comment by mfeurer** [comment]
Please keep this open to remind us to add a way to automatically associate configurations in the ensemble with configurations in `cv_results_`.

**Comment by mfeurer** [comment]
Closing this as the 2nd question is the same as in #1116. I copied the answer to there.
